File: packages/attachment_engine/attachment_engine_ios/ios/attachment_engine_ios/Sources/attachment_engine_ios/AttachmentEnginePlugin.swift
// Copyright 2026 DHC Tech
// Use of this source code is governed by an MIT-style license that can be
// found in the LICENSE file.

import Flutter
import UIKit

/// AttachmentEnginePlugin: registers one focused channel/platform-view per
/// capability, replacing the third-party plugins previously used (pdfx,
/// video_player, just_audio, share_plus, open_filex, dio, path_provider)
/// with hand-written native implementations. Webview embedding is handled
/// by the official `webview_flutter` package instead, at the app-facing
/// layer — not through this plugin. Mirrors `AttachmentEnginePlugin.kt` on
/// Android: identical channel/method names so the Dart side is
/// platform-agnostic.
public class AttachmentEnginePlugin: NSObject, FlutterPlugin {
  private var legacyChannel: FlutterMethodChannel!
  private var messenger: FlutterBinaryMessenger!

  private let pathsChannel = PathsChannel()
  private let pdfChannel = PdfChannel()
  private let downloadChannel = DownloadChannel()
  private let shareChannel = ShareChannel()
  private let openChannel = OpenChannel()
  private let officePreviewChannel = OfficePreviewChannel()
  private let audioChannel = AudioChannel()
  private let videoChannel = VideoChannel()

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = AttachmentEnginePlugin()
    let messenger = registrar.messenger()
    instance.messenger = messenger

    instance.legacyChannel = FlutterMethodChannel(
      name: "attachment_engine", binaryMessenger: messenger)
    registrar.addMethodCallDelegate(instance, channel: instance.legacyChannel)

    instance.pathsChannel.register(with: messenger)
    instance.pdfChannel.register(with: messenger)
    instance.downloadChannel.register(with: messenger)
    instance.shareChannel.register(with: messenger)
    instance.openChannel.register(with: messenger)
    instance.officePreviewChannel.register(with: messenger)
    instance.audioChannel.register(with: messenger)
    instance.videoChannel.register(with: messenger)

    registrar.register(
      VideoPlatformViewFactory(videoChannel: instance.videoChannel),
      withId: VideoChannel.viewType)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getPlatformVersion":
      result("iOS " + UIDevice.current.systemVersion)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  public func detachFromEngine(for registrar: FlutterPluginRegistrar) {
    pathsChannel.unregister(with: messenger)
    pdfChannel.unregister(with: messenger)
    downloadChannel.unregister(with: messenger)
    shareChannel.unregister(with: messenger)
    openChannel.unregister(with: messenger)
    officePreviewChannel.unregister(with: messenger)
    audioChannel.unregister(with: messenger)
    videoChannel.unregister(with: messenger)
  }
}

File: packages/attachment_engine/attachment_engine_android/android/src/main/kotlin/com/dhc/attachment_engine/AttachmentEnginePlugin.kt
package com.dhc.attachment_engine

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/**
 * AttachmentEnginePlugin: registers one Pigeon-generated `HostApi` (or, for
 * the audio/video/download event streams, one hand-written `EventChannel`
 * — see `attachment_engine_platform_interface/pigeons/messages.dart`) plus
 * one platform-view factory, per capability. Replaces the third-party
 * plugins previously used (pdfx, video_player, just_audio, share_plus,
 * open_filex, dio, path_provider) with hand-written native implementations
 * wired through Pigeon codegen for the method-call surface. Webview
 * embedding is handled by the official `webview_flutter` package instead,
 * at the app-facing layer — not through this plugin.
 */
class AttachmentEnginePlugin :
    FlutterPlugin,
    MethodCallHandler {
    private lateinit var legacyChannel: MethodChannel
    private lateinit var messenger: BinaryMessenger

    private lateinit var pathsChannel: PathsChannel
    private lateinit var pdfChannel: PdfChannel
    private lateinit var downloadChannel: DownloadChannel
    private lateinit var shareChannel: ShareChannel
    private lateinit var openChannel: OpenChannel
    private lateinit var audioChannel: AudioChannel
    private lateinit var videoChannel: VideoChannel

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        messenger = flutterPluginBinding.binaryMessenger
        val context = flutterPluginBinding.applicationContext

        legacyChannel = MethodChannel(messenger, "attachment_engine")
        legacyChannel.setMethodCallHandler(this)

        pathsChannel = PathsChannel(context).also { it.register(messenger) }
        pdfChannel = PdfChannel().also { it.register(messenger) }
        downloadChannel = DownloadChannel().also { it.register(messenger) }
        shareChannel = ShareChannel(context).also { it.register(messenger) }
        openChannel = OpenChannel(context).also { it.register(messenger) }
        audioChannel = AudioChannel(messenger).also { it.register() }
        videoChannel = VideoChannel(messenger).also { it.register() }

        flutterPluginBinding.platformViewRegistry.registerViewFactory(
            VideoChannel.VIEW_TYPE,
            VideoPlatformViewFactory(videoChannel),
        )
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        if (call.method == "getPlatformVersion") {
            result.success("Android ${android.os.Build.VERSION.RELEASE}")
        } else {
            result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        legacyChannel.setMethodCallHandler(null)
        pathsChannel.unregister(messenger)
        pdfChannel.unregister(messenger)
        downloadChannel.unregister(messenger)
        shareChannel.unregister(messenger)
        openChannel.unregister(messenger)
        audioChannel.unregister()
        videoChannel.unregister()
    }
}
